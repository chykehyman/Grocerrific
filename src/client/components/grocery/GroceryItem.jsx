import React from 'react';
import { shape, string, bool } from 'prop-types';

const propTypes = {
  item: shape({
    name: string.isRequired,
    purchaseStatus: bool.isRequired
  }).isRequired
};

const GroceryItem = ({ item: { name, purchaseStatus } }) => {
  let listWrapper = 'list-unstyled';
  if (purchaseStatus) {
    listWrapper += ' strike-through';
  }
  return (
    <tr>
      <td>&nbsp;</td>
      <td className={listWrapper}>{name}</td>
      <td>
        <button
          type="button"
          className="btn btn-primary"
        >
          Buy
        </button>
      </td>
      <td>
        <button
          type="button"
          className="btn btn-outline-danger btn-sm">
          Delete
        </button>
      </td>
    </tr>
  );
};

GroceryItem.propTypes = propTypes;

export default GroceryItem;
