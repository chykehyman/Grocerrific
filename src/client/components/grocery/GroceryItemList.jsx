import React from 'react';
import { array, func, string } from 'prop-types';

import GroceryItem from './GroceryItem';

const propTypes = {
  items: array.isRequired,
  selectedId: string.isRequired,
  action: string.isRequired,
  deleteItem: func.isRequired,
  purchaseOrDropItem: func.isRequired
};

const renderNoItemsFound = () => (
  <div className="no-items">
    <i className="fa fa-exclamation-triangle fa-3x pb-3 d-block" />
    <p className="lead">There are no items in the store</p>
  </div>
);

const renderList = (items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem) => (
  <table className="table table-hover">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Name</th>
        <th>Status</th>
        <th>Action</th>
      </tr>
    </thead>
    <tbody>
      {items.map(item => (
        <GroceryItem
          key={item._id}
          item={item}
          isProcessing={isProcessing}
          selectedId={selectedId}
          action={action}
          deleteItem={deleteItem}
          purchaseOrDropItem={purchaseOrDropItem} />
      ))}
    </tbody>
  </table>
);

const GroceryItemList = ({
  items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem
}) => {
  if (items.length === 0) {
    return renderNoItemsFound();
  }
  return renderList(items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem);
};

GroceryItemList.propTypes = propTypes;

export default GroceryItemList;
