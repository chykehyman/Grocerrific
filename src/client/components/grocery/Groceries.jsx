import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import GroceryItemList from './GroceryItemList';

const propTypes = {};

const defaultProps = {};

export class Groceries extends Component {
  groceryItems = [
    { _id: 1, name: 'Tomatoes', purchaseStatus: false },
    { _id: 2, name: 'Peanut Butter', purchaseStatus: true },
    { _id: 3, name: 'Carrots', purchaseStatus: true },
    { _id: 4, name: 'Pinnapples', purchaseStatus: false }
  ];

  render() {
    return (
      <Fragment>
        <GroceryItemList items={this.groceryItems} />
      </Fragment>
    );
  }
}

Groceries.propTypes = propTypes;
Groceries.defaultProps = defaultProps;

export default Groceries;
