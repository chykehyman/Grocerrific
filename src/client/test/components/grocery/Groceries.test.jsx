import { Groceries } from '../../../components/grocery/Groceries';

describe('<Groceries /> component', () => {
  const setup = () => {
    const props = {
      groceries: {
        isFetching: false,
        isProcessing: false,
        groceries: []
      },
      actions: {
        fetchAllGroceries: jest.fn(),
        addGrocery: jest.fn(() => Promise.resolve()),
        deleteGrocery: jest.fn(),
        purchaseOrDropGrocery: jest.fn()
      }
    };

    const state = {
      name: '',
      errors: {},
      selectedId: '',
      action: ''
    };

    return {
      shallowWrapper: shallow(<Groceries {...props} />),
      props,
      state
    };
  };

  const event = {
    target: {
      name: 'tomato'
    },
    preventDefault: jest.fn(),
    persist: jest.fn()
  };


  it('should render without crashing', () => {
    const { shallowWrapper } = setup();

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should display <Loader /> component when `isFetching` is set to true', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.setProps({
      ...props,
      groceries: {
        ...props.groceries,
        isFetching: true
      }
    });

    expect(shallowWrapper.find('MDSpinner')).toHaveLength(1);
  });

  it('should display `Adding...` as button value when `isProcessing` is set to true', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.setProps({
      ...props,
      groceries: {
        ...props.groceries,
        isProcessing: true
      }
    });

    expect(shallowWrapper.find('button').text()).toEqual('Adding...');
  });

  it('invokes `componentDidMount` when mounted', () => {
    const { shallowWrapper } = setup();
    const cdmSpy = jest.spyOn(Groceries.prototype, 'componentDidMount');

    shallowWrapper.instance().componentDidMount();
    expect(cdmSpy).toHaveBeenCalled();
  });

  it('invokes handleOnChange method', () => {
    const { shallowWrapper } = setup();
    const handleOnChangeSpy = jest
      .spyOn(shallowWrapper.instance(), 'handleOnChange');

    shallowWrapper.instance().handleOnChange(event);

    expect(handleOnChangeSpy).toHaveBeenCalled();
  });

  it('invokes handleOnFocus method', () => {
    const { shallowWrapper } = setup();
    const handleOnFocusSpy = jest.spyOn(shallowWrapper.instance(), 'handleOnFocus');

    shallowWrapper.instance().handleOnFocus(event);

    expect(handleOnFocusSpy).toHaveBeenCalled();
  });

  it('invokes handleOnSubmit method', () => {
    const { shallowWrapper } = setup();
    const handleOnSubmitSpy = jest.spyOn(shallowWrapper.instance(), 'handleOnSubmit');
    shallowWrapper.instance().handleOnSubmit(event);

    expect(handleOnSubmitSpy).toHaveBeenCalled();
  });

  it('should throw submit errors when name field is invalid', () => {
    const { shallowWrapper, state } = setup();

    const buttonClick = () => shallowWrapper.find('button').first().simulate('click', event);

    buttonClick();
    expect(shallowWrapper.instance().state.error).toEqual({
      name: 'A minimum of 3 characters is required'
    });

    shallowWrapper.setState({
      ...state, name: '123'
    });

    buttonClick();
    expect(shallowWrapper.instance().state.error).toEqual({
      name: 'field must only contain letters'
    });
  });

  it('should submit item when all validations are met', () => {
    const { shallowWrapper, state } = setup();

    shallowWrapper.setState({
      ...state, name: 'Lemon'
    });

    shallowWrapper.find('button').first().simulate('click', event);

    expect(shallowWrapper.instance().state.error).toEqual({});
  });

  it('should invoke handleOnDelete method', () => {
    const { shallowWrapper } = setup();
    const handleOnDeleteSpy = jest.spyOn(shallowWrapper.instance(), 'handleOnDelete');
    shallowWrapper.instance().handleOnDelete(event);

    expect(handleOnDeleteSpy).toHaveBeenCalled();
  });

  it('should invoke handleOnPurchaseOrDrop method', () => {
    const { shallowWrapper } = setup();
    const handleOnPurchaseOrDropSpy = jest.spyOn(shallowWrapper.instance(), 'handleOnPurchaseOrDrop');
    shallowWrapper.instance().handleOnPurchaseOrDrop(event);

    expect(handleOnPurchaseOrDropSpy).toHaveBeenCalled();
  });
});
