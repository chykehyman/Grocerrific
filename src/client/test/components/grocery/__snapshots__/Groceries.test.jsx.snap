// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Groceries /> component should render without crashing 1`] = `
<Fragment>
  <div
    className="input-group mb-3 mt-4 ml-0 pl-0 col-md-5"
  >
    <input
      autoComplete="off"
      className="form-control"
      name="name"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="enter item name"
      type="text"
      value=""
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add Item
      </button>
    </div>
  </div>
  <GroceryItemList
    action=""
    deleteItem={[Function]}
    isProcessing={false}
    items={Array []}
    purchaseOrDropItem={[Function]}
    selectedId=""
  />
</Fragment>
`;
